### PR TITLE
Add config limit checks for image optimizer

### PR DESCRIPTION
**Pull request by ijjk**
This adds checks to ensure that less than 50 domain and size items are configured and no sizes are less than 1 or greater than 10,000 

x-ref: https://github.com/vercel/next.js/issues/18122

### CODE REVIEW

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Server Mode</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | ijjk/next.js img/add-limit-checks | Change |
| - | - | - | - |
| buildDuration | 13s | 13.1s | ⚠️ +64ms |
| nodeModulesSize | 91.3 MB | 91.3 MB | ⚠️ +1.35 kB |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | ijjk/next.js img/add-limit-checks | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.484 | 2.482 | 0 |
| / avg req/sec | 1006.26 | 1007.06 | +0.8 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.309 | 1.297 | -0.01 |
| /error-in-render avg req/sec | 1909.74 | 1927.52 | +17.78 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js img/add-limit-checks | Change |
| - | - | - | - |
| 677f882d2ed8..574c.js gzip | 11.2 kB | 11.2 kB | ✓ |
| framework.HASH.js gzip | 39 kB | 39 kB | ✓ |
| main-3dd37a4..54de.js gzip | 7.34 kB | 7.34 kB | ✓ |
| webpack-e067..f178.js gzip | 751 B | 751 B | ✓ |
| Overall change | 58.3 kB | 58.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js img/add-limit-checks | Change |
| - | - | - | - |
| 677f882d2ed8..dule.js gzip | 7.02 kB | 7.02 kB | ✓ |
| framework.HA..dule.js gzip | 39 kB | 39 kB | ✓ |
| main-082aff9..dule.js gzip | 6.32 kB | 6.32 kB | ✓ |
| webpack-07c5..dule.js gzip | 751 B | 751 B | ✓ |
| Overall change | 53.1 kB | 53.1 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js img/add-limit-checks | Change |
| - | - | - | - |
| polyfills-4b..e242.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js img/add-limit-checks | Change |
| - | - | - | - |
| _app-9a0b9e1..b37e.js gzip | 1.28 kB | 1.28 kB | ✓ |
| _error-ed1b0..8fbd.js gzip | 3.44 kB | 3.44 kB | ✓ |
| hooks-89731c..c609.js gzip | 887 B | 887 B | ✓ |
| index-17468f..5d83.js gzip | 227 B | 227 B | ✓ |
| link-41ceea0..6b62.js gzip | 1.35 kB | 1.35 kB | ✓ |
| routerDirect..924c.js gzip | 284 B | 284 B | ✓ |
| withRouter-7..c13d.js gzip | 284 B | 284 B | ✓ |
| Overall change | 7.75 kB | 7.75 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages Modern</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js img/add-limit-checks | Change |
| - | - | - | - |
| _app-75d3a82..dule.js gzip | 625 B | 625 B | ✓ |
| _error-4469a..dule.js gzip | 2.29 kB | 2.29 kB | ✓ |
| hooks-cbf13f..dule.js gzip | 387 B | 387 B | ✓ |
| index-b9a643..dule.js gzip | 226 B | 226 B | ✓ |
| link-2d41a89..dule.js gzip | 1.31 kB | 1.31 kB | ✓ |
| routerDirect..dule.js gzip | 284 B | 284 B | ✓ |
| withRouter-f..dule.js gzip | 282 B | 282 B | ✓ |
| Overall change | 5.4 kB | 5.4 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js img/add-limit-checks | Change |
| - | - | - | - |
| _buildManifest.js gzip | 323 B | 323 B | ✓ |
| _buildManife..dule.js gzip | 329 B | 329 B | ✓ |
| Overall change | 652 B | 652 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js img/add-limit-checks | Change |
| - | - | - | - |
| index.html gzip | 1 kB | 1 kB | ✓ |
| link.html gzip | 1.01 kB | 1.01 kB | ✓ |
| withRouter.html gzip | 994 B | 994 B | ✓ |
| Overall change | 3 kB | 3 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Serverless Mode</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | ijjk/next.js img/add-limit-checks | Change |
| - | - | - | - |
| buildDuration | 14.5s | 14.7s | ⚠️ +202ms |
| nodeModulesSize | 91.3 MB | 91.3 MB | ⚠️ +1.35 kB |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js img/add-limit-checks | Change |
| - | - | - | - |
| 677f882d2ed8..574c.js gzip | 11.2 kB | 11.2 kB | ✓ |
| framework.HASH.js gzip | 39 kB | 39 kB | ✓ |
| main-3dd37a4..54de.js gzip | 7.34 kB | 7.34 kB | ✓ |
| webpack-e067..f178.js gzip | 751 B | 751 B | ✓ |
| Overall change | 58.3 kB | 58.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js img/add-limit-checks | Change |
| - | - | - | - |
| 677f882d2ed8..dule.js gzip | 7.02 kB | 7.02 kB | ✓ |
| framework.HA..dule.js gzip | 39 kB | 39 kB | ✓ |
| main-082aff9..dule.js gzip | 6.32 kB | 6.32 kB | ✓ |
| webpack-07c5..dule.js gzip | 751 B | 751 B | ✓ |
| Overall change | 53.1 kB | 53.1 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js img/add-limit-checks | Change |
| - | - | - | - |
| polyfills-4b..e242.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js img/add-limit-checks | Change |
| - | - | - | - |
| _app-9a0b9e1..b37e.js gzip | 1.28 kB | 1.28 kB | ✓ |
| _error-ed1b0..8fbd.js gzip | 3.44 kB | 3.44 kB | ✓ |
| hooks-89731c..c609.js gzip | 887 B | 887 B | ✓ |
| index-17468f..5d83.js gzip | 227 B | 227 B | ✓ |
| link-41ceea0..6b62.js gzip | 1.35 kB | 1.35 kB | ✓ |
| routerDirect..924c.js gzip | 284 B | 284 B | ✓ |
| withRouter-7..c13d.js gzip | 284 B | 284 B | ✓ |
| Overall change | 7.75 kB | 7.75 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages Modern</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js img/add-limit-checks | Change |
| - | - | - | - |
| _app-75d3a82..dule.js gzip | 625 B | 625 B | ✓ |
| _error-4469a..dule.js gzip | 2.29 kB | 2.29 kB | ✓ |
| hooks-cbf13f..dule.js gzip | 387 B | 387 B | ✓ |
| index-b9a643..dule.js gzip | 226 B | 226 B | ✓ |
| link-2d41a89..dule.js gzip | 1.31 kB | 1.31 kB | ✓ |
| routerDirect..dule.js gzip | 284 B | 284 B | ✓ |
| withRouter-f..dule.js gzip | 282 B | 282 B | ✓ |
| Overall change | 5.4 kB | 5.4 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js img/add-limit-checks | Change |
| - | - | - | - |
| _buildManifest.js gzip | 323 B | 323 B | ✓ |
| _buildManife..dule.js gzip | 329 B | 329 B | ✓ |
| Overall change | 652 B | 652 B | ✓ |

</details>

<details>
<summary><strong>Serverless bundles</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js img/add-limit-checks | Change |
| - | - | - | - |
| _error.js | 1.06 MB | 1.06 MB | ✓ |
| 404.html | 4.73 kB | 4.73 kB | ✓ |
| hooks.html | 3.92 kB | 3.92 kB | ✓ |
| index.js | 1.06 MB | 1.06 MB | ✓ |
| link.js | 1.1 MB | 1.1 MB | ✓ |
| routerDirect.js | 1.1 MB | 1.1 MB | ✓ |
| withRouter.js | 1.1 MB | 1.1 MB | ✓ |
| Overall change | 5.42 MB | 5.42 MB | ✓ |

</details>

</details>
Commit: c9eaf7017de6be7621db97f605d8b58e259cb4ab

File: packages/next/next-server/server/config.ts [hunk 1]
```diff
@@ -252,10 +259,20 @@ function assignDefaults(userConfig: { [key: string]: any }) {
...
-      const invalid = images.sizes.filter((d: unknown) => typeof d !== 'number')
+
+      if (images.sizes.length > 50) {
+        throw new Error(
+          `Specified images.sizes exceeds length of 50, received length (${images.sizes.length}), please reduce the length of the array to continue`
+        )
+      }
+
+      const invalid = images.sizes.filter((d: unknown) => {
+        return typeof d !== 'number' || d < 0 || d > 10000
```
Comment (styfle) [inline packages/next/next-server/server/config.ts:270]: Sharp requires an int greater than zero

```
Uncaught Error: Expected positive integer for width but received 0 of type number
    at Object.invalidParameterError (/Users/styfle/Code/vercel/next.js/node_modules/sharp/lib/is.js:101:10)
    at Sharp.resize (/Users/styfle/Code/vercel/next.js/node_modules/sharp/lib/resize.js:214:16)
```

```suggestion
        return typeof d !== 'number' || d < 1 || d > 10000
```

File: packages/next/next-server/server/config.ts [hunk 2]
```diff
@@ -252,10 +259,20 @@ function assignDefaults(userConfig: { [key: string]: any }) {
...
+      })
+
       if (invalid.length > 0) {
         throw new Error(
-          `Specified images.sizes should be an Array of numbers received invalid values (${invalid.join(
+          `Specified images.sizes should be an Array of numbers that are between 0 and 10,000, received invalid values (${invalid.join(
```
Comment (styfle) [inline packages/next/next-server/server/config.ts:275]:
```suggestion
          `Specified images.sizes should be an Array of numbers that are between 1 and 10000, received invalid values (${invalid.join(
```

File: test/integration/image-optimizer/test/index.test.js [hunk 3]
```diff
@@ -321,6 +322,83 @@ function runTests({ w, isDev, domains }) {
...
+        onStderr(msg) {
+          stderr += msg || ''
+        },
+      })
+      await waitFor(1000)
+      await killApp(app).catch(() => {})
+      await nextConfig.restore()
+
+      expect(stderr).toContain(
+        'Specified images.sizes should be an Array of numbers that are between 0 and 10,000, received invalid values (-1, 12000)'
```
Comment (styfle) [inline test/integration/image-optimizer/test/index.test.js:397]:
```suggestion
        'Specified images.sizes should be an Array of numbers that are between 1 and 10000, received invalid values (0, 12000)'
```

File: test/integration/image-optimizer/test/index.test.js [hunk 4]
```diff
@@ -321,6 +322,83 @@ function runTests({ w, isDev, domains }) {
...
+        'Specified images.sizes exceeds length of 50, received length (51), please reduce the length of the array to continue'
+      )
+    })
+
+    it('should error when sizes contains invalid sizes', async () => {
+      await nextConfig.replace(
+        '{ /* replaceme */ }',
+        JSON.stringify({
+          images: {
+            sizes: [-1, 12000, 64, 128, 256],
```
Comment (styfle) [inline test/integration/image-optimizer/test/index.test.js:381]:
```suggestion
            sizes: [0, 12000, 64, 128, 256],
```

**Review by styfle — APPROVED** [review]

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Server Mode</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | ijjk/next.js img/add-limit-checks | Change |
| - | - | - | - |
| buildDuration | 14.5s | 15.2s | ⚠️ +632ms |
| nodeModulesSize | 91.3 MB | 91.3 MB | ⚠️ +1.35 kB |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | ijjk/next.js img/add-limit-checks | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.913 | 2.85 | -0.06 |
| / avg req/sec | 858.29 | 877.18 | +18.89 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.812 | 1.767 | -0.05 |
| /error-in-render avg req/sec | 1379.52 | 1414.77 | +35.25 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js img/add-limit-checks | Change |
| - | - | - | - |
| 677f882d2ed8..574c.js gzip | 11.2 kB | 11.2 kB | ✓ |
| framework.HASH.js gzip | 39 kB | 39 kB | ✓ |
| main-3dd37a4..54de.js gzip | 7.34 kB | 7.34 kB | ✓ |
| webpack-e067..f178.js gzip | 751 B | 751 B | ✓ |
| Overall change | 58.3 kB | 58.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js img/add-limit-checks | Change |
| - | - | - | - |
| 677f882d2ed8..dule.js gzip | 7.02 kB | 7.02 kB | ✓ |
| framework.HA..dule.js gzip | 39 kB | 39 kB | ✓ |
| main-082aff9..dule.js gzip | 6.32 kB | 6.32 kB | ✓ |
| webpack-07c5..dule.js gzip | 751 B | 751 B | ✓ |
| Overall change | 53.1 kB | 53.1 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js img/add-limit-checks | Change |
| - | - | - | - |
| polyfills-4b..e242.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js img/add-limit-checks | Change |
| - | - | - | - |
| _app-9a0b9e1..b37e.js gzip | 1.28 kB | 1.28 kB | ✓ |
| _error-ed1b0..8fbd.js gzip | 3.44 kB | 3.44 kB | ✓ |
| hooks-89731c..c609.js gzip | 887 B | 887 B | ✓ |
| index-17468f..5d83.js gzip | 227 B | 227 B | ✓ |
| link-41ceea0..6b62.js gzip | 1.35 kB | 1.35 kB | ✓ |
| routerDirect..924c.js gzip | 284 B | 284 B | ✓ |
| withRouter-7..c13d.js gzip | 284 B | 284 B | ✓ |
| Overall change | 7.75 kB | 7.75 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages Modern</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js img/add-limit-checks | Change |
| - | - | - | - |
| _app-75d3a82..dule.js gzip | 625 B | 625 B | ✓ |
| _error-4469a..dule.js gzip | 2.29 kB | 2.29 kB | ✓ |
| hooks-cbf13f..dule.js gzip | 387 B | 387 B | ✓ |
| index-b9a643..dule.js gzip | 226 B | 226 B | ✓ |
| link-2d41a89..dule.js gzip | 1.31 kB | 1.31 kB | ✓ |
| routerDirect..dule.js gzip | 284 B | 284 B | ✓ |
| withRouter-f..dule.js gzip | 282 B | 282 B | ✓ |
| Overall change | 5.4 kB | 5.4 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js img/add-limit-checks | Change |
| - | - | - | - |
| _buildManifest.js gzip | 323 B | 323 B | ✓ |
| _buildManife..dule.js gzip | 329 B | 329 B | ✓ |
| Overall change | 652 B | 652 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js img/add-limit-checks | Change |
| - | - | - | - |
| index.html gzip | 1 kB | 1 kB | ✓ |
| link.html gzip | 1.01 kB | 1.01 kB | ✓ |
| withRouter.html gzip | 994 B | 994 B | ✓ |
| Overall change | 3 kB | 3 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Serverless Mode</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | ijjk/next.js img/add-limit-checks | Change |
| - | - | - | - |
| buildDuration | 16.3s | 16.1s | -166ms |
| nodeModulesSize | 91.3 MB | 91.3 MB | ⚠️ +1.35 kB |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js img/add-limit-checks | Change |
| - | - | - | - |
| 677f882d2ed8..574c.js gzip | 11.2 kB | 11.2 kB | ✓ |
| framework.HASH.js gzip | 39 kB | 39 kB | ✓ |
| main-3dd37a4..54de.js gzip | 7.34 kB | 7.34 kB | ✓ |
| webpack-e067..f178.js gzip | 751 B | 751 B | ✓ |
| Overall change | 58.3 kB | 58.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js img/add-limit-checks | Change |
| - | - | - | - |
| 677f882d2ed8..dule.js gzip | 7.02 kB | 7.02 kB | ✓ |
| framework.HA..dule.js gzip | 39 kB | 39 kB | ✓ |
| main-082aff9..dule.js gzip | 6.32 kB | 6.32 kB | ✓ |
| webpack-07c5..dule.js gzip | 751 B | 751 B | ✓ |
| Overall change | 53.1 kB | 53.1 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js img/add-limit-checks | Change |
| - | - | - | - |
| polyfills-4b..e242.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js img/add-limit-checks | Change |
| - | - | - | - |
| _app-9a0b9e1..b37e.js gzip | 1.28 kB | 1.28 kB | ✓ |
| _error-ed1b0..8fbd.js gzip | 3.44 kB | 3.44 kB | ✓ |
| hooks-89731c..c609.js gzip | 887 B | 887 B | ✓ |
| index-17468f..5d83.js gzip | 227 B | 227 B | ✓ |
| link-41ceea0..6b62.js gzip | 1.35 kB | 1.35 kB | ✓ |
| routerDirect..924c.js gzip | 284 B | 284 B | ✓ |
| withRouter-7..c13d.js gzip | 284 B | 284 B | ✓ |
| Overall change | 7.75 kB | 7.75 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages Modern</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js img/add-limit-checks | Change |
| - | - | - | - |
| _app-75d3a82..dule.js gzip | 625 B | 625 B | ✓ |
| _error-4469a..dule.js gzip | 2.29 kB | 2.29 kB | ✓ |
| hooks-cbf13f..dule.js gzip | 387 B | 387 B | ✓ |
| index-b9a643..dule.js gzip | 226 B | 226 B | ✓ |
| link-2d41a89..dule.js gzip | 1.31 kB | 1.31 kB | ✓ |
| routerDirect..dule.js gzip | 284 B | 284 B | ✓ |
| withRouter-f..dule.js gzip | 282 B | 282 B | ✓ |
| Overall change | 5.4 kB | 5.4 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js img/add-limit-checks | Change |
| - | - | - | - |
| _buildManifest.js gzip | 323 B | 323 B | ✓ |
| _buildManife..dule.js gzip | 329 B | 329 B | ✓ |
| Overall change | 652 B | 652 B | ✓ |

</details>

<details>
<summary><strong>Serverless bundles</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js img/add-limit-checks | Change |
| - | - | - | - |
| _error.js | 1.06 MB | 1.06 MB | ✓ |
| 404.html | 4.73 kB | 4.73 kB | ✓ |
| hooks.html | 3.92 kB | 3.92 kB | ✓ |
| index.js | 1.06 MB | 1.06 MB | ✓ |
| link.js | 1.1 MB | 1.1 MB | ✓ |
| routerDirect.js | 1.1 MB | 1.1 MB | ✓ |
| withRouter.js | 1.1 MB | 1.1 MB | ✓ |
| Overall change | 5.42 MB | 5.42 MB | ✓ |

</details>

</details>
Commit: a080c1385b2fbd90c1cedc2d65662d45ed806b66

**Pull request merged**